garak/generators/openai.py

```diff
@@ -257,6 +261,11 @@ class OpenAIGenerator(OpenAICompatible):
     active = True
     generator_family_name = "OpenAI"
 
+    # remove uri as it is not overridable in this class.
+    DEFAULT_PARAMS = {
+        k: val for k, val in OpenAICompatible.DEFAULT_PARAMS.items() if k != "uri"
+    }
+
```

Comment:
this is confusing given the triggering use-case. we should distinguish between generator for OpenAI API Compatible endpoints, and OpenAI Native endpoints hosted by OpenAI - recommend two distinct classes; perhaps one called openai.Compatible, and another called openai.Native, which is the default generator in the module, that implements this `uri` locking

Reply:
I think `openai.Compatible` makes sense but I'd suggest leaving `openai.OpenAIGenerator` as-is. We could also factor `OpenAICompatible` out to another class (maybe to the base class?) in lieu of keeping it in the `openai` generator spec.

Reply:
ah, and then `openai.OpenAIReasoning` for the o1 stuff? OK

Reply:
Do we want to support `base.OpenAICompatible` as a valid `--model_type`? I land on a "no" there.

@erickgalinkin, are you thinking we should push the class down to `base.py` and then have `openai.Compatible` just expose the class, this would come with a requirement for import of `openai` and `backoff` in `base.py`:

``` Python
from garak.generators.base import OpenAICompatible
class Compatible(OpenAICompatible):
    ENV_VAR = "OpenAICompatible_API_KEY".upper()

    active = True
    supports_multiple_generations = True
    generator_family_name = "OpenAICompatible"
```

Depending on the answer above, this seems like a discussion focused on needing a standard in naming expectations for generator classes. There is currently a mix of including the `model_type` value in the classname or not examples:

```
generators: litellm 🌟
generators: litellm.LiteLLMGenerator
```
```
generators: nim 🌟
generators: nim.NVOpenAIChat
generators: nim.NVOpenAICompletion
generators: nim.Vision
```
```
generators: huggingface 🌟
generators: huggingface.ConversationalPipeline
generators: huggingface.InferenceAPI
generators: huggingface.InferenceEndpoint
generators: huggingface.LLaVA
generators: huggingface.Model
generators: huggingface.OptimumPipeline
generators: huggingface.Pipeline
```
```
generators: octo 🌟
generators: octo.InferenceEndpoint
generators: octo.OctoGenerator
```
```
generators: function 🌟
generators: function.Multiple
generators: function.Single
```

I don't see any consistency to say what should be the preferred convention.

Reply:
> Do we want to support `base.OpenAICompatible` as a valid `--model_type`? I land on a "no" there.

Agree, the `base` is confusing - and items in `generators.base` are not functional as-is, iirc; no need to break that pattern.


> Depending on the answer above, this seems like a discussion focused on needing a standard in naming expectations for generator classes. There is currently a mix of including the `model_type` value in the classname or not examples:

Agree. I disprefer repetition. I am missing a word for when the module's default generator is fine named by just the module, though, which is where we often see this.

Reifying:
* avoid repeating module name in generator class name
* strongly avoid repeating module name in non-default generator class name
* avoid including redundant information in class name
* have class name hint at modalities accepted (e.g. Chat, Vision, Completion)

Looking at the examples hinted:

<hr/>

These aren't exposed directly to users, but are overly verbose:
> ```
> generators: litellm.LiteLLMGenerator
> ```
> ```
> generators: octo 🌟
> generators: octo.OctoGenerator
> ```
-- maybe these could become "Chat" or "Completion" or "Text"

<hr/>

These could do with a tidyup:
> ```
> generators: nim 🌟
> generators: nim.NVOpenAIChat
> generators: nim.NVOpenAICompletion
> ```
-- I'm not sure if the NV is redundant, but I expect third-party chat NIMs to continue using this API, and let's wait for a counterexample before working out a general solution. Beyond scope for this PR but could be e.g.
> ```
> generators: nim.Chat
> generators: nim.Completion
> ```

<hr/>

These look pretty good to me, no repetition, and they generally describe the name of the interface using within the context:
> ```
> generators: huggingface 🌟
> generators: huggingface.ConversationalPipeline
> generators: huggingface.InferenceAPI
> generators: huggingface.InferenceEndpoint
> generators: huggingface.LLaVA
> generators: huggingface.Model
> generators: huggingface.OptimumPipeline
> generators: huggingface.Pipeline
> ```
> ```
> generators: function 🌟
> generators: function.Multiple
> generators: function.Single
> ```
> ```
> generators: octo.InferenceEndpoint
> ```
> ```
> generators: nim.Vision
> ```

-- one might rename `huggingface.LLaVA` to `huggingface.VisionLLaVA` for the modality constraint above, if LLaVa isn't considered clue enough.

<hr/>

> I'd suggest leaving openai.OpenAIGenerator as-is. 

@erickgalinkin can I ask what the vibes are like behind this leaning

Reply:
now #1026 has landed, is this ready to be updated?

Reply:
Let's sync on a consensus for naming and I can expand this to add a single new migration based on #1026 landing.

Reply:
Defer rename to future revision, this can land with current probe names.